contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py

```diff
@@ -43,6 +45,16 @@ class Checkstyle(LintTaskMixin, Task):
     FlakeCheckSubsystem,
   )
 
+  @memoized_classproperty
+  def _python_3_compatible_dists(cls):
+    _VERSIONS = ["3.0", "3.1", "3.2", "3.3", "3.4", "3.5", "3.6", "3.7", "3.8", "3.9"]
```

Comment:
This is making me cry. We're right back to all the problems if the initial PR.

A few things are needed to make this bulletproof:
1. We don't need to care about CPython, stackless, pypy, etc here, so best not to track
2. What about "==3.4.5",">3.4,<3.5" etc?

Given this, there are only 2 robust options I can see.
1. Drop `pkg_resources.{Distribution,Requirement}`, and leverage `packaging.{requirements,specifiers,version}` to parse these things fully and then get down to the (operator, version) level to implement a robust test of whether a constraint includes some version of python 3.
2. Whitelist constraints instead.

Option 2 should be much simpler and - I think - get you what you need. The idea would be to whitelist the value of the `python-setup.interpreter_constraints` option taking that as the default for the repo and which currently hardcode defaults to `['>=2.7,<3']`. Lint any target with constraints in that set only by default. Provide a bool option here on the task to allow all - with no mention or care of "3". Alternatively - if a bool is not enough, make the option a list of interpreter constraints to whitelist that defaults to the value of `python-setup.interpreter_constraints`  and can be appended, replaced or blanked as the magic value to say "all".

Reply:
What do you think of a third option:
run minimal interpreter selection and see if the minimum interpreter is a 3 interpreter? That way we can get compatibility checking at the cost of setting up the interpreter cache for filters:
```
allowed_interpreters = set(interpreter_cache.setup(filters=filters))
if not allowed_interpreters:
  raise TaskError('No valid interpreters found for targets: {}\n(filters: {})'
                              .format(targets, filters))
interpreter = min(allowed_interpreters)
```

I like option 2 but afaict it puts us in the same position as the original code because we need constraint intersection logic per option 1.

Reply:
We spoke offline, but I don't think intersection is needed. The idea is to treat interpreter constraints as a finite set of labels and whitelist from that set. The motivating assumption being real world repos have 2-3 interpreter constraints. Canonically 2, the default pants hardcoded ">=2.7,<3" and maybe an additional ">=3.4,<4". In this example you'd selectively add ">=3.4,<4" to the new whitelist option. For robustness you still need to parse things using `packaging`, but you only whitelist exact matches.